File: src/bun.js/webcore/blob/Store.zig
const Store = @This();

data: Data,

mime_type: MimeType = .none,
ref_count: std.atomic.Value(u32) = .init(1),
is_all_ascii: ?bool = null,
allocator: std.mem.Allocator,

pub const new = bun.TrivialNew(@This());

pub fn memoryCost(this: *const Store) usize {
    return if (this.hasOneRef()) @sizeOf(@This()) + switch (this.data) {
        .bytes => this.data.bytes.len,
        .file => 0,
        .s3 => |s3| s3.estimatedSize(),
    } else 0;
}

pub fn getPath(this: *const Store) ?[]const u8 {
    return switch (this.data) {
        .bytes => |*bytes| if (bytes.stored_name.len > 0) bytes.stored_name.slice() else null,
        .file => |*file| if (file.pathlike == .path) file.pathlike.path.slice() else null,
        .s3 => |*s3| s3.pathlike.slice(),
    };
}

pub fn size(this: *const Store) SizeType {
    return switch (this.data) {
        .bytes => this.data.bytes.len,
        .s3, .file => Blob.max_size,
    };
}

pub const Map = std.HashMap(u64, *jsc.WebCore.Blob.Store, bun.IdentityContext(u64), 80);

pub const Data = union(enum) {
    bytes: Bytes,
    file: File,
    s3: S3,
};

pub fn ref(this: *Store) void {
    const old = this.ref_count.fetchAdd(1, .monotonic);
    assert(old > 0);
}

pub fn hasOneRef(this: *const Store) bool {
    return this.ref_count.load(.monotonic) == 1;
}

/// Caller is responsible for derefing the Store.
pub fn toAnyBlob(this: *Store) ?Blob.Any {
    if (this.hasOneRef()) {
        if (this.data == .bytes) {
            return .{ .InternalBlob = this.data.bytes.toInternalBlob() };
        }
    }

    return null;
}

pub fn external(ptr: ?*anyopaque, _: ?*anyopaque, _: usize) callconv(.c) void {
    if (ptr == null) return;
    var this = bun.cast(*Store, ptr);
    this.deref();
}
pub fn initS3WithReferencedCredentials(pathlike: node.PathLike, mime_type: ?MimeType, credentials: *bun.S3.S3Credentials, allocator: std.mem.Allocator) !*Store {
    var path = pathlike;
    // this actually protects/refs the pathlike
    path.toThreadSafe();

    const store = Blob.Store.new(.{
        .data = .{
            .s3 = S3.initWithReferencedCredentials(
                path,
                mime_type orelse brk: {
                    const sliced = path.slice();
                    if (sliced.len > 0) {
                        var extname = std.fs.path.extension(sliced);
                        extname = std.mem.trim(u8, extname, ".");
                        if (MimeType.byExtensionNoDefault(extname)) |mime| {
                            break :brk mime;
                        }
                    }
                    break :brk null;
                },
                credentials,
            ),
        },
        .allocator = allocator,
        .ref_count = std.atomic.Value(u32).init(1),
    });
    return store;
}

pub fn initS3(pathlike: node.PathLike, mime_type: ?MimeType, credentials: bun.S3.S3Credentials, allocator: std.mem.Allocator) !*Store {
    var path = pathlike;
    // this actually protects/refs the pathlike
    path.toThreadSafe();

    const store = Blob.Store.new(.{
        .data = .{
            .s3 = S3.init(
                path,
                mime_type orelse brk: {
                    const sliced = path.slice();
                    if (sliced.len > 0) {
                        var extname = std.fs.path.extension(sliced);
                        extname = std.mem.trim(u8, extname, ".");
                        if (MimeType.byExtensionNoDefault(extname)) |mime| {
                            break :brk mime;
                        }
                    }
                    break :brk null;
                },
                credentials,
            ),
        },
        .allocator = allocator,
        .ref_count = std.atomic.Value(u32).init(1),
    });
    return store;
}
pub fn initFile(pathlike: jsc.Node.PathOrFileDescriptor, mime_type: ?MimeType, allocator: std.mem.Allocator) !*Store {
    const store = Blob.Store.new(.{
        .data = .{
            .file = File.init(
                pathlike,
                mime_type orelse brk: {
                    if (pathlike == .path) {
                        const sliced = pathlike.path.slice();
                        if (sliced.len > 0) {
                            var extname = std.fs.path.extension(sliced);
                            extname = std.mem.trim(u8, extname, ".");
                            if (MimeType.byExtensionNoDefault(extname)) |mime| {
                                break :brk mime;
                            }
                        }
                    }

                    break :brk null;
                },
            ),
        },
        .allocator = allocator,
        .ref_count = std.atomic.Value(u32).init(1),
    });
    return store;
}

/// Takes ownership of `bytes`, which must have been allocated with `allocator`.
pub fn init(bytes: []u8, allocator: std.mem.Allocator) *Store {
    const store = Blob.Store.new(.{
        .data = .{
            .bytes = Bytes.init(bytes, allocator),
        },
        .allocator = allocator,
        .ref_count = .init(1),
    });
    return store;
}

pub fn sharedView(this: Store) []u8 {
    if (this.data == .bytes)
        return this.data.bytes.slice();

    return &[_]u8{};
}

pub fn deref(this: *Blob.Store) void {
    const old = this.ref_count.fetchSub(1, .monotonic);
    assert(old >= 1);
    if (old == 1) {
        this.deinit();
    }
}

pub fn deinit(this: *Blob.Store) void {
    const allocator = this.allocator;

    switch (this.data) {
        .bytes => |*bytes| {
            bytes.deinit();
        },
        .file => |file| {
            if (file.pathlike == .path) {
                if (file.pathlike.path == .string) {
                    allocator.free(@constCast(file.pathlike.path.slice()));
                } else {
                    file.pathlike.path.deinit();
                }
            }
        },
        .s3 => |*s3| {
            s3.deinit(allocator);
        },
    }

    bun.destroy(this);
}

pub const SerializeTag = enum(u8) {
    file = 0,
    bytes = 1,
    empty = 2,
};

pub fn serialize(this: *Store, comptime Writer: type, writer: Writer) !void {
    switch (this.data) {
        .file => |file| {
            const pathlike_tag: jsc.Node.PathOrFileDescriptor.SerializeTag = if (file.pathlike == .fd) .fd else .path;
            try writer.writeInt(u8, @intFromEnum(pathlike_tag), .little);

            switch (file.pathlike) {
                .fd => |fd| {
                    try writer.writeStruct(fd);
                },
                .path => |path| {
                    const path_slice = path.slice();
                    try writer.writeInt(u32, @as(u32, @truncate(path_slice.len)), .little);
                    try writer.writeAll(path_slice);
                },
            }
        },
        .s3 => |s3| {
            const pathlike_tag: jsc.Node.PathOrFileDescriptor.SerializeTag = .path;
            try writer.writeInt(u8, @intFromEnum(pathlike_tag), .little);

            const path_slice = s3.pathlike.slice();
            try writer.writeInt(u32, @as(u32, @truncate(path_slice.len)), .little);
            try writer.writeAll(path_slice);
        },
        .bytes => |bytes| {
            const slice = bytes.slice();
            try writer.writeInt(u32, @truncate(slice.len), .little);
            try writer.writeAll(slice);

            try writer.writeInt(u32, @truncate(bytes.stored_name.slice().len), .little);
            try writer.writeAll(bytes.stored_name.slice());
        },
    }
}

pub fn fromArrayList(list: std.ArrayListUnmanaged(u8), allocator: std.mem.Allocator) !*Blob.Store {
    return try Blob.Store.init(list.items, allocator);
}

/// A blob store that references a file on disk.
pub const File = struct {
    pathlike: jsc.Node.PathOrFileDescriptor,
    mime_type: MimeType = MimeType.other,
    is_atty: ?bool = null,
    mode: bun.Mode = 0,
    seekable: ?bool = null,
    max_size: SizeType = Blob.max_size,
    // milliseconds since ECMAScript epoch
    last_modified: jsc.JSTimeType = jsc.init_timestamp,

    pub fn unlink(this: *const File, globalThis: *JSGlobalObject) bun.JSError!JSValue {
        return switch (this.pathlike) {
            .path => |path_like| jsc.Node.fs.Async.unlink.create(globalThis, undefined, .{
                .path = .{
                    .encoded_slice = switch (path_like) {
                        .encoded_slice => |slice| try slice.toOwned(bun.default_allocator),
                        else => try jsc.ZigString.init(path_like.slice()).toSliceClone(bun.default_allocator),
                    },
                },
            }, globalThis.bunVM()),
            .fd => jsc.JSPromise.resolvedPromiseValue(globalThis, globalThis.createInvalidArgs("Is not possible to unlink a file descriptor", .{})),
        };
    }
    pub fn isSeekable(this: *const File) ?bool {
        if (this.seekable) |seekable| {
            return seekable;
        }

        if (this.mode != 0) {
            return bun.isRegularFile(this.mode);
        }

        return null;
    }

    pub fn init(pathlike: jsc.Node.PathOrFileDescriptor, mime_type: ?MimeType) File {
        return .{ .pathlike = pathlike, .mime_type = mime_type orelse MimeType.other };
    }
};

/// An S3 Blob Store
pub const S3 = struct {
    pathlike: node.PathLike,
    mime_type: MimeType = .other,
    credentials: ?*S3Credentials,
    options: bun.S3.MultiPartUploadOptions = .{},
    acl: ?bun.S3.ACL = null,
    storage_class: ?bun.S3.StorageClass = null,
    request_payer: bool = false,

    pub fn isSeekable(_: *const @This()) ?bool {
        return true;
    }

    pub fn getCredentials(this: *const @This()) *S3Credentials {
        bun.assert(this.credentials != null);
        return this.credentials.?;
    }

    pub fn getCredentialsWithOptions(this: *const @This(), options: ?JSValue, globalObject: *JSGlobalObject) bun.JSError!bun.S3.S3CredentialsWithOptions {
        return S3Credentials.getCredentialsWithOptions(this.getCredentials().*, this.options, options, this.acl, this.storage_class, this.request_payer, globalObject);
    }

    pub fn path(this: *@This()) []const u8 {
        var path_name = bun.URL.parse(this.pathlike.slice()).s3Path();
        // normalize start and ending
        if (strings.endsWith(path_name, "/")) {
            path_name = path_name[0..path_name.len];
        } else if (strings.endsWith(path_name, "\\")) {
            path_name = path_name[0 .. path_name.len - 1];
        }
        if (strings.startsWith(path_name, "/")) {
            path_name = path_name[1..];
        } else if (strings.startsWith(path_name, "\\")) {
            path_name = path_name[1..];
        }
        return path_name;
    }

    pub fn unlink(this: *@This(), store: *Store, globalThis: *JSGlobalObject, extra_options: ?JSValue) bun.JSError!JSValue {
        const Wrapper = struct {
            promise: jsc.JSPromise.Strong,
            store: *Store,
            global: *JSGlobalObject,

            pub const new = bun.TrivialNew(@This());

            pub fn resolve(result: bun.S3.S3DeleteResult, opaque_self: *anyopaque) bun.JSTerminated!void {
                const self: *@This() = @ptrCast(@alignCast(opaque_self));
                defer self.deinit();
                const globalObject = self.global;
                switch (result) {
                    .success => {
                        try self.promise.resolve(globalObject, .true);
                    },
                    .not_found, .failure => |err| {
                        try self.promise.reject(globalObject, err.toJS(globalObject, self.store.getPath()));
                    },
                }
            }

            fn deinit(wrap: *@This()) void {
                wrap.store.deref();
                wrap.promise.deinit();
                bun.destroy(wrap);
            }
        };
        const promise = jsc.JSPromise.Strong.init(globalThis);
        const value = promise.value();
        const proxy_url = globalThis.bunVM().transpiler.env.getHttpProxy(true, null, null);
        const proxy = if (proxy_url) |url| url.href else null;
        var aws_options = try this.getCredentialsWithOptions(extra_options, globalThis);
        defer aws_options.deinit();
        store.ref();

        try bun.S3.delete(&aws_options.credentials, this.path(), @ptrCast(&Wrapper.resolve), Wrapper.new(.{
            .promise = promise,
            .store = store, // store is needed in case of not found error
            .global = globalThis,
        }), proxy, aws_options.request_payer);

        return value;
    }

    pub fn listObjects(this: *@This(), store: *Store, globalThis: *JSGlobalObject, listOptions: JSValue, extra_options: ?JSValue) bun.JSError!JSValue {
        if (!listOptions.isEmptyOrUndefinedOrNull() and !listOptions.isObject()) {
            return globalThis.throwInvalidArguments("S3Client.listObjects() needs a S3ListObjectsOption as it's first argument", .{});
        }

        const Wrapper = struct {
            promise: jsc.JSPromise.Strong,
            store: *Store,
            resolvedlistOptions: bun.S3.S3ListObjectsOptions,
            global: *JSGlobalObject,

            pub fn resolve(result: bun.S3.S3ListObjectsResult, opaque_self: *anyopaque) bun.JSTerminated!void {
                const self: *@This() = @ptrCast(@alignCast(opaque_self));
                defer self.deinit();
                const globalObject = self.global;

                switch (result) {
                    .success => |list_result| {
                        defer list_result.deinit();
                        const list_result_js = list_result.toJS(globalObject) catch return self.promise.reject(globalObject, error.JSError);
                        try self.promise.resolve(globalObject, list_result_js);
                    },

                    inline .not_found, .failure => |err| {
                        try self.promise.reject(globalObject, err.toJS(globalObject, self.store.getPath()));
                    },
                }
            }

            fn deinit(self: *@This()) void {
                self.store.deref();
                self.promise.deinit();
                self.resolvedlistOptions.deinit();
                self.destroy();
            }

            pub inline fn destroy(self: *@This()) void {
                bun.destroy(self);
            }
        };

        const promise = jsc.JSPromise.Strong.init(globalThis);
        const value = promise.value();
        const proxy_url = globalThis.bunVM().transpiler.env.getHttpProxy(true, null, null);
        const proxy = if (proxy_url) |url| url.href else null;
        var aws_options = try this.getCredentialsWithOptions(extra_options, globalThis);
        defer aws_options.deinit();

        const options = try bun.S3.getListObjectsOptionsFromJS(globalThis, listOptions);
        store.ref();

        try bun.S3.listObjects(&aws_options.credentials, options, @ptrCast(&Wrapper.resolve), bun.new(Wrapper, .{
            .promise = promise,
            .store = store, // store is needed in case of not found error
            .resolvedlistOptions = options,
            .global = globalThis,
        }), proxy);

        return value;
    }

    pub fn initWithReferencedCredentials(pathlike: node.PathLike, mime_type: ?MimeType, credentials: *S3Credentials) S3 {
        credentials.ref();
        return .{
            .credentials = credentials,
            .pathlike = pathlike,
            .mime_type = mime_type orelse MimeType.other,
        };
    }
    pub fn init(pathlike: node.PathLike, mime_type: ?MimeType, credentials: S3Credentials) S3 {
        return .{
            .credentials = credentials.dupe(),
            .pathlike = pathlike,
            .mime_type = mime_type orelse MimeType.other,
        };
    }
    pub fn estimatedSize(this: *const @This()) usize {
        return this.pathlike.estimatedSize() + if (this.credentials) |credentials| credentials.estimatedSize() else 0;
    }

    pub fn deinit(this: *@This(), allocator: std.mem.Allocator) void {
        if (this.pathlike == .string) {
            allocator.free(@constCast(this.pathlike.slice()));
        } else {
            this.pathlike.deinit();
        }
        this.pathlike = .{
            .string = bun.PathString.empty,
        };
        if (this.credentials) |credentials| {
            credentials.deref();
            this.credentials = null;
        }
    }

    const S3Credentials = bun.S3.S3Credentials;
};

pub const Bytes = struct {
    ptr: ?[*]u8 = undefined,
    len: SizeType = 0,
    cap: SizeType = 0,
    allocator: std.mem.Allocator,

    /// Used by standalone module graph and the File constructor
    stored_name: bun.PathString = bun.PathString.empty,

    /// Takes ownership of `bytes`, which must have been allocated with
    /// `allocator`.
    pub fn init(bytes: []u8, allocator: std.mem.Allocator) Bytes {
        return .{
            .ptr = bytes.ptr,
            .len = @as(SizeType, @truncate(bytes.len)),
            .cap = @as(SizeType, @truncate(bytes.len)),
            .allocator = allocator,
        };
    }
    pub fn initEmptyWithName(name: bun.PathString, allocator: std.mem.Allocator) Bytes {
        return .{
            .ptr = null,
            .len = 0,
            .cap = 0,
            .allocator = allocator,
            .stored_name = name,
        };
    }

    pub fn fromArrayList(list: std.ArrayListUnmanaged(u8), allocator: std.mem.Allocator) !*Bytes {
        return Bytes.init(list.items, allocator);
    }

    pub fn toInternalBlob(this: *Bytes) Blob.Internal {
        const ptr = this.ptr orelse return .{
            .bytes = std.array_list.Managed(u8){
                .items = &.{},
                .capacity = 0,
                .allocator = this.allocator,
            },
        };

        const result: Blob.Internal = .{
            .bytes = .{
                .items = ptr[0..this.len],
                .capacity = this.cap,
                .allocator = this.allocator,
            },
        };

        this.allocator = bun.default_allocator;
        this.len = 0;
        this.cap = 0;
        return result;
    }
    pub fn slice(this: Bytes) []u8 {
        if (this.ptr) |ptr| {
            return ptr[0..this.len];
        }
        return "";
    }

    pub fn allocatedSlice(this: Bytes) []u8 {
        if (this.ptr) |ptr| {
            return ptr[0..this.cap];
        }
        return "";
    }

    pub fn deinit(this: *Bytes) void {
        bun.default_allocator.free(this.stored_name.slice());
        if (this.ptr) |ptr| {
            this.allocator.free(ptr[0..this.cap]);
        }
        this.ptr = null;
        this.len = 0;
        this.cap = 0;
    }

    pub fn asArrayList(this: Bytes) std.ArrayListUnmanaged(u8) {
        return this.asArrayListLeak();
    }

    pub fn asArrayListLeak(this: Bytes) std.ArrayListUnmanaged(u8) {
        return .{
            .items = this.ptr[0..this.len],
            .capacity = this.cap,
        };
    }
};

const std = @import("std");

const bun = @import("bun");
const assert = bun.assert;
const strings = bun.strings;
const webcore = bun.webcore;
const MimeType = bun.http.MimeType;
const node = bun.api.node;

const jsc = bun.jsc;
const JSGlobalObject = jsc.JSGlobalObject;
const JSValue = jsc.JSValue;

const Blob = webcore.Blob;
const SizeType = Blob.SizeType;

File: src/runtime/webcore/blob/Store.rs
//! `Blob.Store` — backing storage variants for `webcore::Blob`.
//!
//! LAYERING: the data types (`Store`/`RefPtr<Store>`/`Data`/`Bytes`/`File`/`S3`)
//! are the **single nominal definitions** in `bun_jsc::webcore_types::store`;
//! this module re-exports them and layers the `bun_runtime`-tier behaviour
//! (S3 I/O, async file ops, structured-clone serialize) via extension traits.

use crate::node::fs as node_fs;
use crate::node::types::PathOrFileDescriptorSerializeTag;
use crate::webcore::jsc::{JSGlobalObject, JSPromise, JSValue, JsResult};
use crate::webcore::node_types::{PathLike, PathOrFileDescriptor};
use crate::webcore::s3::client as s3_client;
use crate::webcore::s3::client::S3ErrorJsc as _;
use crate::webcore::s3::client::{
    S3Credentials, S3CredentialsWithOptions, S3DeleteResult, S3ListObjectsResult,
};
use bun_core::strings;
use bun_http_types::MimeType::MimeType;
use bun_ptr::RefPtr;
use bun_url::URL;

// ──────────────────────────────────────────────────────────────────────────
// Re-export the canonical data types from `bun_jsc`.
// ──────────────────────────────────────────────────────────────────────────

pub use bun_jsc::webcore_types::store::{
    Bytes, Data, DataTag, File, IsAllAscii, S3, SerializeTag, Store,
};

// ──────────────────────────────────────────────────────────────────────────
// Extension traits — `bun_runtime`-tier behaviour layered on the `bun_jsc`
// data types. Inherent data-only methods (`size`/`shared_view`/`ref_`/`deref`/
// `init`/…) live on the `bun_jsc` types directly.
// ──────────────────────────────────────────────────────────────────────────

pub trait StoreExt {
    fn to_any_blob(&mut self) -> Option<super::Any>;
    fn init_s3(
        pathlike: PathLike<'static>,
        mime_type: Option<MimeType>,
        credentials: S3Credentials,
    ) -> Result<RefPtr<Store>, crate::Error>
    where
        Self: Sized;
    fn init_file(
        pathlike: PathOrFileDescriptor<'static>,
        mime_type: Option<MimeType>,
    ) -> Result<RefPtr<Store>, crate::Error>
    where
        Self: Sized;
    fn serialize(&self, writer: &mut impl bun_io::Write) -> Result<(), crate::Error>;
}

pub trait S3Ext {
    fn get_credentials_with_options(
        &self,
        options: Option<JSValue>,
        global_object: &JSGlobalObject,
    ) -> JsResult<S3CredentialsWithOptions>;
    /// `store` is the heap `Store` that owns `self` (`self == &store.data.S3`);
    /// the request keeps its own ref on it.
    fn unlink(
        &self,
        store: &RefPtr<Store>,
        global_this: &JSGlobalObject,
        extra_options: Option<JSValue>,
    ) -> JsResult<JSValue>;
    /// See `unlink`.
    fn list_objects(
        &self,
        store: &RefPtr<Store>,
        global_this: &JSGlobalObject,
        list_options: JSValue,
        extra_options: Option<JSValue>,
    ) -> JsResult<JSValue>;
}

pub trait FileExt {
    fn unlink(&self, global_this: &JSGlobalObject) -> JsResult<JSValue>;
}

pub trait BytesExt {
    fn to_internal_blob(&mut self) -> super::Internal;
}

/// Shared mime-sniffing fallback for the `init_*` constructors below: derive a
/// `MimeType` from the path's extension, returning `None` for empty paths or
/// unknown extensions.
#[inline]
fn mime_from_path_ext(sliced: &[u8]) -> Option<MimeType> {
    if sliced.is_empty() {
        return None;
    }
    let ext = strings::trim(bun_paths::extension(sliced), b".");
    bun_http_types::MimeType::by_extension_no_default(ext)
}

impl StoreExt for Store {
    /// Caller is responsible for derefing the Store.
    fn to_any_blob(&mut self) -> Option<super::Any> {
        if self.has_one_ref() {
            if let Data::Bytes(bytes) = &mut self.data {
                return Some(super::Any::InternalBlob(bytes.to_internal_blob()));
            }
        }

        None
    }

    fn init_s3(
        pathlike: PathLike<'static>,
        mime_type: Option<MimeType>,
        credentials: S3Credentials,
    ) -> Result<RefPtr<Store>, crate::Error> {
        let path = pathlike.thread_isolated_copy();

        // Compute the extension-derived fallback before moving `path` into the
        // Store so we don't need to clone the owned PathLike.
        let mime_type = mime_type.or_else(|| mime_from_path_ext(path.slice()));

        Ok(RefPtr::new(Store {
            data: Data::S3(S3::init(path, mime_type, credentials)),
            mime_type: bun_http_types::MimeType::NONE,
            ref_count: bun_ptr::ThreadSafeRefCount::init(),
            is_all_ascii: IsAllAscii::default(),
        }))
    }

    fn init_file(
        pathlike: PathOrFileDescriptor<'static>,
        mime_type: Option<MimeType>,
    ) -> Result<RefPtr<Store>, crate::Error> {
        // Compute the extension-derived fallback before moving `pathlike` into
        // the Store so we don't need to clone the owned PathOrFileDescriptor.
        let mime_type = mime_type.or_else(|| match &pathlike {
            PathOrFileDescriptor::Path(path) => mime_from_path_ext(path.slice()),
            PathOrFileDescriptor::Fd(_) => None,
        });

        Ok(RefPtr::new(Store {
            data: Data::File(File::init(pathlike, mime_type)),
            mime_type: bun_http_types::MimeType::NONE,
            ref_count: bun_ptr::ThreadSafeRefCount::init(),
            is_all_ascii: IsAllAscii::default(),
        }))
    }

    fn serialize(&self, writer: &mut impl bun_io::Write) -> Result<(), crate::Error> {
        match &self.data {
            Data::File(file) => {
                let pathlike_tag: PathOrFileDescriptorSerializeTag =
                    if matches!(file.pathlike, PathOrFileDescriptor::Fd(_)) {
                        PathOrFileDescriptorSerializeTag::Fd
                    } else {
                        PathOrFileDescriptorSerializeTag::Path
                    };
                writer.write_int_le::<u8>(pathlike_tag as u8)?;

                match &file.pathlike {
                    PathOrFileDescriptor::Fd(fd) => {
                        // Write the raw bytes of the FD wrapper. `bun_sys::Fd` is
                        // `#[repr(transparent)]` over an integer (`i32` posix /
                        // `u64` windows), so its native-endian byte image is
                        // exactly the inner field's `to_ne_bytes()`.
                        writer.write_all(&fd.0.to_ne_bytes())?;
                    }
                    PathOrFileDescriptor::Path(path) => {
                        let path_slice = path.slice();
                        writer.write_int_le::<u32>(path_slice.len() as u32)?;
                        writer.write_all(path_slice)?;
                    }
                }
            }
            Data::S3(s3) => {
                let pathlike_tag = PathOrFileDescriptorSerializeTag::Path;
                writer.write_int_le::<u8>(pathlike_tag as u8)?;

                let path_slice = s3.pathlike.slice();
                writer.write_int_le::<u32>(path_slice.len() as u32)?;
                writer.write_all(path_slice)?;
            }
            Data::Bytes(bytes) => {
                let slice = bytes.slice();
                writer.write_int_le::<u32>(slice.len() as u32)?;
                writer.write_all(slice)?;

                writer.write_int_le::<u32>(bytes.stored_name.len() as u32)?;
                writer.write_all(&bytes.stored_name)?;
            }
        }
        Ok(())
    }
}

impl FileExt for File {
    fn unlink(&self, global_this: &JSGlobalObject) -> JsResult<JSValue> {
        match &self.pathlike {
            PathOrFileDescriptor::Path(path_like) => {
                // The `*Binding` arg is unused in `AsyncFSTask::create`.
                let binding = node_fs::Binding::default();
                Ok(node_fs::async_::Unlink::create(
                    global_this,
                    &binding,
                    node_fs::args::Unlink::owned(path_like.slice().to_vec()),
                    global_this.bun_vm().as_mut(),
                ))
            }
            PathOrFileDescriptor::Fd(_) => Ok(JSPromise::resolved_promise_value(
                global_this,
                // `JSGlobalObject::create_invalid_args` lives in the still-gated
                // `JSGlobalObject.rs`; `ERR_INVALID_ARG_TYPE` (lib.rs) is the
                // same `ErrorCode::INVALID_ARG_TYPE.fmt(...)` body.
                global_this.ERR_INVALID_ARG_TYPE(format_args!(
                    "Is not possible to unlink a file descriptor"
                )),
            )),
        }
    }
}

impl S3Ext for S3 {
    fn get_credentials_with_options(
        &self,
        options: Option<JSValue>,
        global_object: &JSGlobalObject,
    ) -> JsResult<S3CredentialsWithOptions> {
        // The associated fn (surfaced via `S3CredentialsExt` in `webcore/S3Client.rs`)
        // takes `&S3Credentials` instead of by-value because `S3Credentials` carries a
        // private intrusive ref-count and cannot be struct-copied; the impl deep-copies
        // internally.
        use crate::webcore::s3_client::S3CredentialsExt as _;
        S3Credentials::get_credentials_with_options(
            self.get_credentials(),
            self.options,
            options,
            self.acl,
            self.storage_class,
            self.request_payer,
            global_object,
        )
    }

    fn unlink(
        &self,
        store: &RefPtr<Store>,
        global_this: &JSGlobalObject,
        extra_options: Option<JSValue>,
    ) -> JsResult<JSValue> {
        let mut promise = bun_jsc::JSPromiseStrong::init(global_this);
        let value = promise.value();
        let global = bun_ptr::BackRef::new(global_this);
        let store = store.clone();
        let resolve = move |result: S3DeleteResult<'_>| -> JsResult<()> {
            let global_object = global.get();
            match result {
                S3DeleteResult::Success => {
                    promise.resolve(global_object, JSValue::TRUE)?;
                }
                S3DeleteResult::NotFound(err) | S3DeleteResult::Failure(err) => {
                    // Split borrows: `reject` takes `&mut promise`, so
                    // compute the error (which reads `promise.get()`) first.
                    let err_val =
                        err.to_js_with_async_stack(global_object, store.get_path(), promise.get());
                    promise.reject(global_object, err_val)?;
                }
            }
            Ok(())
        };

        // `Transpiler::env_mut` is the safe accessor for the process-singleton
        // dotenv loader (never null once the VM is initialised).
        let proxy_url: Option<URL<'_>> = global_this
            .bun_vm()
            .as_mut()
            .transpiler
            .env_mut()
            .get_http_proxy(true, None, None);
        let proxy = proxy_url.as_ref().map(|url| url.href);
        let aws_options = self.get_credentials_with_options(extra_options, global_this)?;
        // `defer aws_options.deinit()` → Drop handles it.

        s3_client::delete(
            &aws_options.credentials,
            self.path(),
            Box::new(resolve),
            proxy,
            aws_options.request_payer,
        )?;

        Ok(value)
    }

    fn list_objects(
        &self,
        store: &RefPtr<Store>,
        global_this: &JSGlobalObject,
        list_options: JSValue,
        extra_options: Option<JSValue>,
    ) -> JsResult<JSValue> {
        if !list_options.is_empty_or_undefined_or_null() && !list_options.is_object() {
            return Err(global_this.throw_invalid_arguments(format_args!(
                "S3Client.listObjects() needs a S3ListObjectsOption as it's first argument"
            )));
        }

        let mut promise = bun_jsc::JSPromiseStrong::init(global_this);
        let value = promise.value();
        let global = bun_ptr::BackRef::new(global_this);
        let store = store.clone();
        let resolve = move |result: S3ListObjectsResult<'_>| -> JsResult<()> {
            let global_object = global.get();
            match result {
                S3ListObjectsResult::Success(list_result) => {
                    let list_result_js = match list_result.to_js(global_object) {
                        Ok(v) => v,
                        Err(e) => {
                            return promise.reject(global_object, Err(e));
                        }
                    };
                    promise.resolve(global_object, list_result_js)?;
                }

                S3ListObjectsResult::NotFound(err) | S3ListObjectsResult::Failure(err) => {
                    // Split borrows: `reject` takes `&mut promise`, so
                    // compute the error (which reads `promise.get()`) first.
                    let err_val =
                        err.to_js_with_async_stack(global_object, store.get_path(), promise.get());
                    promise.reject(global_object, err_val)?;
                }
            }
            Ok(())
        };

        // `Transpiler::env_mut` is the safe accessor for the process-singleton
        // dotenv loader (never null once the VM is initialised).
        let proxy_url: Option<URL<'_>> = global_this
            .bun_vm()
            .as_mut()
            .transpiler
            .env_mut()
            .get_http_proxy(true, None, None);
        let proxy = proxy_url.as_ref().map(|url| url.href);
        let aws_options = self.get_credentials_with_options(extra_options, global_this)?;
        // `defer aws_options.deinit()` → Drop handles it.

        // Only read synchronously, to build the search-params string.
        let options = s3_client::get_list_objects_options_from_js(global_this, list_options)?;

        s3_client::list_objects(&aws_options.credentials, &options, Box::new(resolve), proxy)?;

        Ok(value)
    }
}

impl BytesExt for Bytes {
    fn to_internal_blob(&mut self) -> super::Internal {
        // `Internal.bytes` is `Vec<u8>` (global allocator): the allocation
        // itself when the storage *is* the global allocator's, otherwise a
        // copy (and the original is freed through its allocator, e.g. munmap).
        super::Internal {
            bytes: self.take_vec(),
            was_string: false,
        }
    }
}
